File: todomvc-distributed-application/src/main/java/de/muspellheim/todomvc/distributed/BackendProxy.java
/*
 * TodoMVC - Distributed Application
 * Copyright (c) 2021 Falko Schumann <falko.schumann@muspellheim.de>
 */

package de.muspellheim.todomvc.distributed;

import de.muspellheim.messages.CommandStatus;
import de.muspellheim.messages.Failure;
import de.muspellheim.messages.HttpCommandStatus;
import de.muspellheim.todomvc.contract.MessageHandling;
import de.muspellheim.todomvc.contract.messages.commands.ClearCompletedCommand;
import de.muspellheim.todomvc.contract.messages.commands.DestroyCommand;
import de.muspellheim.todomvc.contract.messages.commands.EditCommand;
import de.muspellheim.todomvc.contract.messages.commands.NewTodoCommand;
import de.muspellheim.todomvc.contract.messages.commands.ToggleAllCommand;
import de.muspellheim.todomvc.contract.messages.commands.ToggleCommand;
import de.muspellheim.todomvc.contract.messages.queries.TodosQuery;
import de.muspellheim.todomvc.contract.messages.queries.TodosQueryResult;
import java.util.List;

class BackendProxy implements MessageHandling {
  public static final String BACKEND_API_BASE_URL = "http://localhost:8080/api/";

  private final HttpJsonClient client = new HttpJsonClient(BACKEND_API_BASE_URL);

  @Override
  public CommandStatus handle(ClearCompletedCommand command) {
    try {
      return client
          .execute("clear-completed-command", command, HttpCommandStatus.class)
          .commandStatus();
    } catch (Exception e) {
      return new Failure(e.getLocalizedMessage());
    }
  }

  @Override
  public CommandStatus handle(DestroyCommand command) {
    try {
      return client.execute("destroy-command", command, HttpCommandStatus.class).commandStatus();
    } catch (Exception e) {
      return new Failure(e.getLocalizedMessage());
    }
  }

  @Override
  public CommandStatus handle(EditCommand command) {
    try {
      return client.execute("edit-command", command, HttpCommandStatus.class).commandStatus();
    } catch (Exception e) {
      return new Failure(e.getLocalizedMessage());
    }
  }

  @Override
  public CommandStatus handle(NewTodoCommand command) {
    try {
      return client.execute("new-todo-command", command, HttpCommandStatus.class).commandStatus();
    } catch (Exception e) {
      return new Failure(e.getLocalizedMessage());
    }
  }

  @Override
  public CommandStatus handle(ToggleAllCommand command) {
    try {
      return client.execute("toggle-all-command", command, HttpCommandStatus.class).commandStatus();
    } catch (Exception e) {
      return new Failure(e.getLocalizedMessage());
    }
  }

  @Override
  public CommandStatus handle(ToggleCommand command) {
    try {
      return client.execute("toggle-command", command, HttpCommandStatus.class).commandStatus();
    } catch (Exception e) {
      return new Failure(e.getLocalizedMessage());
    }
  }

  @Override
  public TodosQueryResult handle(TodosQuery query) {
    try {
      return client.execute("todos-query", query, TodosQueryResult.class);
    } catch (Exception e) {
      return new TodosQueryResult(List.of());
    }
  }
}

File: todomvc-single-page-application/src/backend/BackendProxy.ts
import { MessageHandling } from '../contract/MessageHandling';
import {
  ClearCompletedCommand,
  CommandStatus,
  DestroyCommand,
  EditCommand,
  NewTodoCommand,
  ToggleAllCommand,
  ToggleCommand,
} from '../contract/messages/commands';
import { TodosQuery, TodosQueryResult } from '../contract/messages/queries';
import { HttpJsonClient } from './HttpJsonClient';

const BACKEND_URL = 'http://localhost:8080/api/';

export class BackendProxy implements MessageHandling {
  private client = new HttpJsonClient(BACKEND_URL);

  async handleClearCompletedCommand(command: ClearCompletedCommand): Promise<CommandStatus> {
    try {
      return await this.client.execute('clear-completed-command', command);
    } catch (error) {
      return { success: false, errorMessage: error };
    }
  }

  async handleDestroyCommand(command: DestroyCommand): Promise<CommandStatus> {
    try {
      return await this.client.execute('destroy-command', command);
    } catch (error) {
      return { success: false, errorMessage: error };
    }
  }

  async handleEditCommand(command: EditCommand): Promise<CommandStatus> {
    try {
      return await this.client.execute('edit-command', command);
    } catch (error) {
      return { success: false, errorMessage: error };
    }
  }

  async handleNewTodoCommand(command: NewTodoCommand): Promise<CommandStatus> {
    try {
      return await this.client.execute('new-todo-command', command);
    } catch (error) {
      return { success: false, errorMessage: error };
    }
  }

  async handleToggleAllCommand(command: ToggleAllCommand): Promise<CommandStatus> {
    try {
      return this.client.execute('toggle-all-command', command);
    } catch (error) {
      return { success: false, errorMessage: error };
    }
  }

  async handleToggleCommand(command: ToggleCommand): Promise<CommandStatus> {
    try {
      return await this.client.execute('toggle-command', command);
    } catch (error) {
      return { success: false, errorMessage: error };
    }
  }

  async handleTodosQuery(query: TodosQuery): Promise<TodosQueryResult> {
    try {
      return await this.client.execute('todos-query', query);
    } catch (error) {
      return { todos: [] };
    }
  }
}
